### PR TITLE
Unique IDs

### PR DESCRIPTION
**Pull request by holgerdell**
- I added a required field `id: string` to `QuestionGenerator`. The idea is that these IDs will be unique and stable among algo-learn instances that we control. This is important so that we can safely store learning data that will not change if we rearrange the generators in the UI. Uniqueness is ensured using unit testing, stability is not ensured in any way (best-effort).
- The URI-based routing to the generator/parameter/seed combinations now uses the generator ids directly and is therefore independent from how the UI chooses to organize the generators.
- Added a new type `QuestionCollection` that allows to define a list of generator lists (e.g., arrange generators by topic in the same way that we've been doing it so far). The front-end is free to change this later if needed.
- `useSkills.ts` attempts to automatically update existing learning data based on the old paths to the new, stable URIs.

To-Do in a future pull request:
- refactor useSkills
- show relevant learning data near each generator in the UI

### CODE REVIEW

File: shared/src/api/QuestionGenerator.ts [hunk 1]
```diff
@@ -195,6 +195,9 @@ export type FreeTextFormatFunction = (
  * example MultipleChoiceQuestion, etc.
  */
 export interface QuestionGenerator {
+  /** Unique and stable identifier of the question generator. */
+  id: string
```
Comment (AbcAeffchen) [inline shared/src/api/QuestionGenerator.ts:199]: This id is used directly in the URLs. So it should be documented that this is the case and maybe what values we expect here. E.g. it cannot contain `/` and some other characters. maybe we want to have them short. But as it looks the currently have a word or an abbreviation and no hash or something like that.

Comment (holgerdell) [reply shared/src/api/QuestionGenerator.ts:203]: This was only documented in the unit test, and I added a comment here to explain the currently rather strict rules.

File: shared/src/api/QuestionRouter.ts [hunk 2]
```diff
@@ -1,59 +1,53 @@
...
+ * Objects of the following type describe a sequence of question collections.
  */
-export type QuestionRoutes = Array<{
-  path: string
-  generator: QuestionGenerator
+export type QuestionCollection = Array<{
+  slug: string
+  name: SingleTranslation
+  contents: Array<QuestionGenerator>
+  image?: URL
```
Comment (Fabianstw) [inline shared/src/api/QuestionRouter.ts:14]: Do we want to keep the images? If yes, I wouldn't make this optional, so its mandatory for every collection to have it's own image.

Comment (holgerdell) [reply shared/src/api/QuestionRouter.ts:14]: Currently, they are not displayed. If image is undefined, there is a default image that would be used instead

Comment (Fabianstw) [reply shared/src/api/QuestionRouter.ts:14]: Yes, currently they are not being displayed, I meant looking into the future. But if a default image can be used it should not matter 👍

**Pull request merged**